x-pack/metricbeat/module/aws/cloudwatch/cloudwatch.go

```diff
@@ -0,0 +1,326 @@
+	}
+
+	return &MetricSet{
+		MetricSet: metricSet,
+	}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(report mb.ReporterV2) error {
+	// Get CloudwatchMetrics Config
+	cloudwatchConfigs := m.CloudwatchMetrics
+	if len(cloudwatchConfigs) == 0 {
+		return errors.New("cloudwatch_metrics in config is missing")
+	}
+	// Get startTime and endTime
+	startTime, endTime, err := aws.GetStartTimeEndTime(m.DurationString)
+	if err != nil {
+		return errors.Wrap(err, "Error ParseDuration")
+	}
+
+	// Get listMetricsTotal and namespaces from configuration
+	listMetricsTotal, namespacesTotal := readCloudwatchConfig(cloudwatchConfigs)
+
+	// Use listMetricsTotal from config
+	svcCloudwatch := cloudwatch.New(*m.MetricSet.AwsConfig)
+	err = createEvents(svcCloudwatch, listMetricsTotal, m.PeriodInSec, startTime, endTime, report)
+	if err != nil {
+		return errors.New("createEvents failed")
+	}
+
+	// Use namespaces from config
+	for _, namespace := range namespacesTotal {
+		for _, regionName := range m.MetricSet.RegionsList {
+			m.MetricSet.AwsConfig.Region = regionName
+			svcCloudwatch := cloudwatch.New(*m.MetricSet.AwsConfig)
+			listMetricsOutput, err := aws.GetListMetricsOutput(namespace, regionName, svcCloudwatch)
+			if err != nil {
+				m.Logger().Errorf(err.Error())
+				report.Error(err)
+				continue
+			}
+
+			if listMetricsOutput == nil || len(listMetricsOutput) == 0 {
+				continue
+			}
+
+			err = createEvents(svcCloudwatch, listMetricsOutput, m.PeriodInSec, startTime, endTime, report)
+			if err != nil {
+				return errors.Wrap(err, "createEvents failed for region "+regionName)
+			}
+		}
+	}
+
+	return nil
+}
+
+func readCloudwatchConfig(cloudwatchConfigs []map[string]interface{}) ([]cloudwatch.Metric, []string) {
+	listMetricsTotal := []cloudwatch.Metric{}
+	namespacesTotal := []string{}
+	for _, cloudwatchConfig := range cloudwatchConfigs {
+		namespace := cloudwatchConfig["namespace"].(string)
+		if cloudwatchConfig["metricname"] != nil {
+			listMetricsOutput := convertConfigToListMetrics(cloudwatchConfig, namespace)
+			listMetricsTotal = append(listMetricsTotal, listMetricsOutput)
+		} else {
+			namespacesTotal = append(namespacesTotal, namespace)
+		}
+	}
+	return listMetricsTotal, namespacesTotal
+}
+
+func constructMetricQueries(listMetricsOutput []cloudwatch.Metric, period int64) []cloudwatch.MetricDataQuery {
+	metricDataQueries := []cloudwatch.MetricDataQuery{}
+	for i, listMetric := range listMetricsOutput {
+		metricDataQuery := createMetricDataQuery(listMetric, i, period)
+		metricDataQueries = append(metricDataQueries, metricDataQuery)
+	}
+	return metricDataQueries
+}
+
+func constructLabel(metric cloudwatch.Metric) string {
+	metricDims := metric.Dimensions
+	metricName := *metric.MetricName
```

Comment:
probably there is no need to create these 2 variables?

Reply:
nope, cleaned! thanks!